initial implementation of kuberc

mpuckett159 opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR seeks to implement the initial alpha version of the kuberc feature that was proposed in [KEP 3104](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/3104-introduce-kuberc). This is a very early stage attempt at this and definitely has a lot of work to be done on it and leaves a lot to be desired, particularly the fact that we're manually parsing out flags instead of using cobra's built in flag functionality. This is intended to be fixed up.

#### Which issue(s) this PR fixes:
N/A

#### Special notes for your reviewer:
This is a very early attempt and will definitely need to be iterated on and improved. Just trying to get something out the door that is at least mostly workable/usable. One hugely notable draw back is that the `--kuberc` flag can not be called in the form of `--kuberc="path/to/kuberc"` and must be called using the form `--kuberc path/to/kuberc`.

Another thing is that I have done nothing to implement the default flags for every command, but that shouldn't be too hard after we get consensus on how the changes I've made.

I haven't done much testing on this yet and am opening this PR as a draft to gather some feedback from the sig-cli chairs on the methodology I'm using and make sure my changes to the YAML config file are acceptable.

This PR also makes use of the github.com/ohler55/ojg library which is what I intend to use to replace the in tree jsonpath library that we have in k/k. This seemed simpler and more straightforward to read/understand in code than doing a bunch of weird recursive look ups.

#### Does this PR introduce a user-facing change?
```release-note
Introduces an alpha version of the kuberc functionality
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP 3104]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/3104-introduce-kuberc
```

mpuckett159 committed:
initial implementation of kuberc

mpuckett159 commented:
I know these are probably automated messages dims is sending out but just for clarity this WIP is for example and discussion regarding the initial kuberc implementation.

dims commented:
> I know these are probably automated messages dims is sending out but just for clarity this WIP is for example and discussion regarding the initial kuberc implementation.

Ack, thanks @mpuckett159 !

mpuckett159 commented:
After some consideration and review of Maciej's PR showing me the less painful way of intercepting command line args, I have some changes I'm going to make to the structure of this, namely the format of the config file.

Below you will see and example of what I want the final config file format to look like. I don't remember why I originally started using the term "overrides" for the default flags part but we can change that whenever, it's relatively inconsequential. I have split out the command from the flags as Maciej did in his PR. This will allow for simpler code on our side, no ugly text parsing because we will have the flags and the commands separate from the get go.

I have changed up the overrides section as well most specifically by not nesting the nested commands within themselves. This makes for a flatter and much easier to traverse config file, I'm not sure why it didn't occur to me that it doesn't matter if the map key has a space in it or not but I got there eventually. The default flag structure is much the same as the alias structure, just minus the command itself. I'm still working on getting this implemented fully so that it makes use of my kuberc file reading functions and such, but hopefully that shouldn't take too much more work. I think that I'm just about there. Thanks for the help and collaboration all.

```apiVersion: v1alpha1
kind: Preferences
command:
  aliases:
    getdbprod:
      command: get pods
      flags:
        - -l what=database
        - --namespace us-2-production
    getdbdev:
      command: get pods
      flags:
        - -l what=database
        - --namespace us-2-development
  overrides:
    apply:
      flags:
        - --server-side=true
    config:
      flags:
        - --kubeconfig "~/.kube/config"
    config set:
      flags:
        - --set-raw-bytes
    config view:
      flags:
        - --raw
    delete:
      flags:
        - --confirm=true
    default:
      flags:
        - --exec-auth-allowlist "/var/kubectl/exec/..."```

mpuckett159 commented:
Note to self the way this works is the aliases get added to the final subcommand in the command string so in my example config file `getdbprod` only gets added as an alias for the subcommand `get->pods` and can not be passed on it's own at the top level. A way around this should be to return the command part of the alias and replace the last word with the alias itself so people don't have to do `kubectl get getdbprod`. Though this may be desirable I dunno, that remains an open question I guess. I guess that should be configurable, maybe default to only being able to be run that way, but add a bool that will allow the alias to be run from the top level like `kubectl getdbprod`. I'll have to bring that up on the meeting and see what people think I guess.

mpuckett159 commented:
Alright the other branch is working pretty well now, command inheritance goes user supplied -> alias -> override. The main missing features at this point are:

1. Set nondefault path for kuberc file
2. Set global flags like `--kubeconfig` and the like. I think I'm just not understanding how that works though, will continue research and solicit some hints on that front.

mpuckett159 committed:
Feature/kuberc config alt (#2)

* Implement hints from maciej and change kuberc file format

* Plumb cmdutil.Factory into kubectl config subcommands

* Fix config subcommands

* fix imports

ardaguclu commented on staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go:
```diff
@@ -37,11 +37,14 @@ import (
 	flagAuthInfoName       = "user"
 	flagContext            = "context"
 	flagNamespace          = "namespace"
+	shortFlagNamespace     = "n"
```
For example, how do we know that we are covering everything in addition to `n` and `s` and how people in the future will know that new shortcuts should be added in here?

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -95,7 +96,10 @@ var defaultConfigFlags = genericclioptions.NewConfigFlags(true).WithDeprecatedPa
 // NewDefaultKubectlCommand creates the `kubectl` command with default arguments
 func NewDefaultKubectlCommand() *cobra.Command {
+	// TODO: add something to check for --kuberc flag and override filepath
+	homeDir, _ := os.UserHomeDir()
```
As you said in the TODO already, this should be the same with how `kubeconfig` handles this.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -323,6 +327,12 @@ func NewKubectlCommand(o KubectlOptions) *cobra.Command {
 	if kubeConfigFlags == nil {
 		kubeConfigFlags = defaultConfigFlags
 	}
+	if os.Getenv("ENABLE_KUBERC") != "" {
```
nitpick: could you please use https://github.com/kubernetes/kubernetes/blob/04675428bbfc9bf7ba4c9e1abfc427b6228069d9/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L428

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -323,6 +327,12 @@ func NewKubectlCommand(o KubectlOptions) *cobra.Command {
+	if os.Getenv("ENABLE_KUBERC") != "" {
+		klog.Infoln("kuberc is enabled!")
+		if o.KubercHandler != nil {
+			o.KubercHandler.InjectOverridesRoot(kubeConfigFlags, o.Arguments)
```
I think this should be below of `kubeConfigFlags.AddFlags(flags)` not above(during my local tests, if it is above the `AddFlags`, defaulting did not work.)

Have you tried only injecting defaults in kuberc file and seen what cobra does when user explicitly passes flag after `AddFlags` executed?

I mean;

```go
kubeConfigFlags.AddFlags(flags)
if cmdutil.KubeRc.IsEnabled() && o.KubercHandler != nil {
o.KubercHandler.InjectOverridesRoot(kubeconfigflags) // kubeconfigflags are set with kuberc settings via flag.Set() function 
}
```

mpuckett159 replied:
I believe that causes a break in the setting of the flags for the subcommands (minus config) but I need to double check.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/cmd.go:
```diff
@@ -432,7 +442,7 @@ func NewKubectlCommand(o KubectlOptions) *cobra.Command {
 
 	cmds.AddCommand(alpha)
-	cmds.AddCommand(cmdconfig.NewCmdConfig(clientcmd.NewDefaultPathOptions(), o.IOStreams))
+	cmds.AddCommand(cmdconfig.NewCmdConfig(f, o.IOStreams))
```
In order to preserve compatibility with other commands, passing factory instead PathOptions makes sense. However, this looks like a behavioral change needs careful consideration and I think, it should be in totally separate PR prior to kuberc?

mpuckett159 replied:
Agreed should definitely be it's own PR.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/kuberc/kuberc.go:
```diff
@@ -0,0 +1,199 @@
+	var preferences Preferences
+	// TODO: (mpuckett159) This probably should be UnmarshalStrict but I'm not sure
+	if err := yaml.Unmarshal(kubercBytes, &preferences); err != nil {
+		klog.Warningf("error unmarshalling the yaml for kuberc")
```
What happens if user has corrupted `kubeconfig`?, I think it returns error and we should also return error and intercept the request. Otherwise, users will file issues about kuberc not working without knowing that their kuberc yaml is invalid.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/kuberc/kuberc.go:
```diff
@@ -0,0 +1,199 @@
+//kind: Preferences
+//command:
+//  aliases:
+//    dbprod:
```
I'd prefer this way instead;
```yaml
aliases:
   get pods:
     ...
```
`dbprod` should be unique and this way more emphasizes on this requirement.

mpuckett159 replied:
So the thing is that the aliases will only be unique to the subcommand itself and will not be unique globally. As far as I can tell cobra doesn't allow global aliasing, so you would be running commands like `k get dbprod` and not `k dbprod`. See [here](https://github.com/spf13/cobra/issues/1725) for more details.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/kuberc/kuberc.go:
```diff
@@ -0,0 +1,199 @@
+	}
+	if flags, found := h.Command.Overrides[strings.Join(args, " ")]; found {
+		klog.V(2).Infof("adding default flags %q for command %q", flags.Flags, strings.Join(args, " "))
+		if err := cmd.Flags().Parse(flags.Flags); err != nil {
```
nice function I did not know

ardaguclu reviewed (commented):
That is a good effort @mpuckett159, thanks.

ardaguclu commented on staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go:
```diff
@@ -124,6 +127,137 @@ type ConfigFlags struct {
+// OverwriteDefaultConfigFlags returns ConfigFlags with default values set when no
+// arg is passed, and the set arg for any passed values. The intent of this is to
+// overwrite the default flags with new defaults from the kuberc config file
+func (f *ConfigFlags) OverwriteDefaultConfigFlags(flags []string, args []string) {
```
What I'd prefer if this is the only way to implement kuberc, don't support global config defaulting if this ^ is the only way :)

mpuckett159 replied:
Yeah I think this PR has highlighted some short comings with how kubectl handles global root level flags as a whole. I think I need to work on fixing that in a separate PR, similar to fixing the config subcommands. I need to investigate that further because, yes, writing this custom parser was janky and not fun.

mpuckett159 committed:
Use FeatureGate properly

mpuckett159 committed:
Hand back proper error for yaml unmarshal error

dims commented:
This work-in-progress PR needs a rebase. please rebase if this PR is still needed for the upcoming release.
